crates/optimism/txpool/src/validator.rs

```diff
@@ -225,14 +209,81 @@ where
         &self,
         transactions: Vec<(TransactionOrigin, Tx)>,
     ) -> Vec<TransactionValidationOutcome<Tx>> {
-        futures_util::future::join_all(
-            transactions.into_iter().map(|(origin, tx)| self.validate_one(origin, tx)),
-        )
+        let transactions = transactions
+            .into_iter()
+            .map(|(origin, tx)| {
+                self.apply_op_checks_no_state(tx)
+                    .and_then(|tx| self.inner.validate_one_no_state(origin, tx))
+                    .map(|tx| (origin, tx))
+            })
+            .collect::<Vec<_>>();
+
+        // bail early if all transactions failed validation without state
+        let some_pass = transactions.iter().any(|res| res.is_ok());
+        if !some_pass {
+            return transactions.into_iter().filter_map(|res| res.err()).collect()
+        }
+
+        // otherwise load state from DB for checks against state
+        let state: Arc<dyn StateProvider> = match self.client().latest() {
+            Ok(s) => s.into(),
+            Err(err) => {
+                return transactions
+                    .into_iter()
+                    .map(|res| match res {
+                        Ok((_, tx)) => {
+                            TransactionValidationOutcome::Error(*tx.hash(), Box::new(err.clone()))
+                        }
+                        Err(already_invalid) => already_invalid,
+                    })
+                    .collect()
+            }
+        };
```

Comment:
we don't need to use `Arc` here atm, since l1 validation can be done before op checks since https://github.com/paradigmxyz/reth/pull/15843. could still be useful to keep the generic state provider instead of constraining it to `Box<dyn StateProvider>` for sake of flexibility